applications/adv_networking_bench/cpp/dpdk_bench_op_rx.h

```diff
@@ -136,91 +139,157 @@ class AdvNetworkingBenchDefaultRxOp : public Operator {
+
+          // TODO: could free CPU packets (segment 0) now
         }
       } else {
+        // Batched: headers and payload to GPU (queue memory regions should be a single GPU segment)
         for (int p = 0; p < adv_net_get_num_pkts(burst); p++) {
+          // Get pointers to payload data on GPU (shifting by header size)
+          // NOTE: currently ordering pointers in the order packets come in. If headers had segment
+          //       ID, the index in h_dev_ptrs_ should use that (instead of aggr_pkts_recv_ + p).
           h_dev_ptrs_[cur_batch_idx_][aggr_pkts_recv_ + p] =
               reinterpret_cast<uint8_t*>(adv_net_get_seg_pkt_ptr(burst, 0, p)) + header_size_.get();
+
+          // Count bytes received
           ttl_bytes_recv_ += adv_net_get_seg_pkt_len(burst, 0, p);
         }
       }
     } else {
+      /* CPU Mode (needs to match if the ANO queue uses no memory regions)
+       * Copy each packet payload in a continuous host-pinned buffer, copy of that larger buffer to
+       * the GPU will occur later (copying each packet to GPU directly would be too expensive).
+       *
+       * NOTE: this assume huge pages memory regions. With host-pinned memory regions, this could be
+       *       skipped, though probably not faster given the higher perf to write to huge pages.
+       */
+
+      // Calculate offset for the ANO burst
+      // NOTE: we could keep track of each packet length and aggregate it instead.
       auto burst_offset = aggr_pkts_recv_ * nom_payload_size_;
+
       for (int p = 0; p < adv_net_get_num_pkts(burst); p++) {
+        // Payload address (UDPIPV4Pkt: + 1 skips the header)
         auto payload_ptr = static_cast<UDPIPV4Pkt*>(adv_net_get_seg_pkt_ptr(burst, 0, p)) + 1;
+        // Payload length (packet length minus header length)
+        // NOTE: this should be equal to nom_payload_size_ as we assume the same length for all
+        //       packets in this sample app
         auto pkt_len = adv_net_get_seg_pkt_len(burst, 0, p);
         auto payload_len = pkt_len - header_size_.get();
 
+        // Copy payload to aggregated CPU buffers now
         memcpy((char*)full_batch_data_h_[cur_batch_idx_] + burst_offset + p * nom_payload_size_,
                payload_ptr,
                payload_len);
 
+        // Count bytes received
         ttl_bytes_recv_ += pkt_len;
+
+        // TODO: could free CPU packets now
       }
     }
 
+    /* For each ANO batch (named burst), we might not want to right away send the packets to the next
+     * operator, but maybe wait for more packets to come in, to make up what we call an "App batch".
+     * While that increases the latency by needing more data to come in to continue, it is useful
+     * in two scenarios:
+     *
+     *   1. CPU mode or HDS mode: reduce CPU memory usage by not holding onto packets that can be freed
+     *      (whether full packet buffers or headers only for HDS)
```

Comment:
Where can I learn more about HDS mode?

Reply:
Current definition is here: https://github.com/nvidia-holoscan/holohub/blob/883997d168604c023c2445a72bcb08ef13923645/applications/adv_networking_bench/cpp/dpdk_bench_op_rx.h#L151

Larger docs revamp with more context will come later, this is minor additions in the meantime.

Reply:
Works for me